lib/index.ts

```diff
@@ -263,9 +263,13 @@ export async function performBuilds(
 		secrets: secretMap,
 		regSecrets: registrySecrets,
 		architecture,
+		experimental,
 	} = await initializeBuildMetadata(tasks, docker, tmpDir);
 
 	const images = await Bluebird.map(tasks, (task: BuildTask) => {
+		if (experimental) {
+			task.dockerOpts = { platform: task.dockerPlatform, ...task.dockerOpts };
```

Comment:
Something I found during CLI testing (`balena build` and `balena deploy --build` commands) is that the `pull: true` field is also needed in `dockerOpts` in order to cause Docker / balenaEngine to re-pull the image for a different platform / architecture, otherwise it will use any locally existing image that matches the image tag even if it is for the wrong architecture. This actually sounds like a Docker bug... I tested it with:
```
$ balena-engine version
Server:
 Engine:
  Version:          18.09.10-dev
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       7cb464a406748016f2df0c31a9851d20456a3d31
  Built:            Tue Mar 10 17:10:13 2020
  OS/Arch:          linux/amd64
  Experimental:     true
```
(Tested on Linux, where `docker` / balenaEngine does not use [binfmt_misc](https://en.wikipedia.org/wiki/Binfmt_misc).)

The `pull: true` option is described in:
* https://docs.docker.com/engine/api/v1.39/#operation/ImageBuild  
`pull  	Attempt to pull the image even if an older image exists locally.`
* https://docs.docker.com/engine/reference/commandline/build/#extended-description  
`docker build --pull`:  
`--pull  	Always attempt to pull a newer version of the image`

To put it another way, without the `pull: true` field in `dockerOpts`:
* `balena build test-multiarch --arch amd64 --deviceType intel-nuc --nocache`  
Assuming no matching base image existed before this command, a base image is correctly pulled for the `linux/amd64`. It works.
* `balena build test-multiarch --arch rpi --deviceType rpi --nocache`  
Running this command after the previous command, a base image already existed (for the `linux/amd64` platform) and therefore no new image is pulled for the `linux/arm/v6` platform. The `linux/amd64` image is used and the build fails with "exec format error". Issuing `docker rmi` command(s) for the base image(s), then re-running `balena build`, causes the `linux/arm/v6` image to be pulled and then `balena build` succeeds.

But I am also not saying that `pull: true` should be always there, as it could be a significant performance issue. I am considering adding a CLI option `balena build --pull` and warning messages advising users to add `pull` if the CLI host platform mismatches the application or device platform.

Reply:
hmm that's messy. I'm not sure how to best deal with this... possibly the cli option + warnings is the best compromise. I also wouldn't want to be too smart and conditionally add the option

Reply:
CLI PR balena-io/balena-cli/pull/2015 has added a `--pull` option to the `balena build` and `balena deploy` commands, so the CLI sets the `pull: true` Docker option when it contacts the Docker daemon directly. However, this does not cover `balena push`, where it would be for the builder to set to the `pull: true` Docker API option.

Reply:
the reason for this would be the emulated builds right... I would actually like to be able to use `binfmt_misc` and specifying `balena build test-multiarch --arch amd64,aarch64`

Reply:
on the other hand... did you try with a v19.03.x release of docker/balenaEngine?

Reply:
I tested with Docker Desktop on macOS:
```
$ docker --version
Docker version 19.03.8, build afacb8b
```
But it never raises any `exec format error` because it always uses `binfmt_misc`, so I couldn't tell whether it was using the right base image or not. :-)   Would `uname -a` report the architecture of the image, or the architecture of the actual processor, when `binfmt_misc` is active?...

Reply:
hmm I can think of some additional experiments I could try, like `balena deploy --build` with Docker 19.03, deploying to an RPi that should fail to execute the image if Docker selected an amd64 image...

Reply:
I have answered some of my own questions:
* `uname -a` reports the architecture of the image, not the architecture of the actual processor when `binfmt_misc` is used for emulation, and therefore can be used to test the selection of base images.
* "Docker version 19.03.8, build afacb8b" (Docker Desktop for macOS) also requires `pull: true` (i.e. the `--pull` option in `balena build --pull` or `balena deploy --build --pull`) in order to refresh / select the correct base image architecture if a previous base image of same tag but different architecture already existed locally. (To me, this feels like a Docker bug, because Docker is effectively happy to use a base image for the wrong architecture.)

For `balena build` and `balena deploy`, I think we can get away with expecting users to add `--pull` in order for the correct base image to be downloaded, provided that a warning message is printed if the app / image architecture doesn't match the host architecture.

However, I am not yet clear about `balena push`, where `pull: true` (if needed) should be added by the balenaCloud builder.

Reply:
@pdcastro the builder always does `pull: true` so we are safe in the cloud. But in the case of the cli, it feels wrong that the user has to pass always `--pull`. On the other hand, I wouldn't do an implicit `--pull` when building/deploying, since is an unexpected behavior

Reply:
although to be honest, I'd like the builder to be consistent with the cli, will ease our lives from the maintenance point of view

Reply:
> But in the case of the cli, it feels wrong that the user has to pass always `--pull`. On the other hand, I wouldn't do an implicit `--pull` when building/deploying, since is an unexpected behavior.

Users don't have to pass always `--pull`, only when they want the effect of checking for base image updates / re-pulling images. :-) I read a comment from Page where he said that `--pull` does not necessarily cause a base image to be re-downloaded: it would only be re-downloaded if Docker detects that it has changed. And yet, testing with Docker Desktop on macOS, adding the `--pull` option causes a delay compatible with re-downloading. I need to double check whether or not a re-download was indeed happening.

This discussion re `--pull` behavior does not affect this `resin-multibuild` PR though. The `--pull` option was added to the CLI a few weeks ago (optional behavior), and "historically" (several years) the CLI does not pass the `pull` option to the Docker daemon. A change to this behavior would indeed be a breaking change.

Reply:
I created https://github.com/balena-os/balena-engine/issues/229 to track this possible bug in the engine